service/history/taskProcessor.go

```diff
@@ -246,9 +249,20 @@ func (t *taskProcessor) processTaskOnce(
 	default:
 	}
 
+	var scopeIdx int
+	var err error
+
+	domainID := task.task.GetDomainID()
+	scope := t.metricsScopeCache.Get(domainID)
+
 	startTime := t.timeSource.Now()
-	scopeIdx, err := task.processor.process(task)
-	scope := t.metricsClient.Scope(scopeIdx).Tagged(t.getDomainTagByID(task.task.GetDomainID()))
+
+	if scope == nil {
+		scopeIdx, err = task.processor.process(task)
```

Comment:
So we only process if scope is nil?

Reply:
The code is a little bit confusing here, `process()` does two things: 1. process the task, 2. return a metric scope ID that should be used the for all the metrics for this task. 

So we always need to process the task, regardless of if the scope exists in the cache or not.